src/nio/events/misc.py

```diff
@@ -114,9 +109,48 @@ class UnknownBadEvent:
     sender_key: Optional[str] = field(default=None, init=False)
     session_id: Optional[str] = field(default=None, init=False)
 
+    def flattened(
+        self,
+        _prefix: str = "",
+        _source: Optional[Dict[str, Any]] = None,
+        _flat: Optional[Dict[str, Any]] = None,
+    ) -> Dict[str, Any]:
+        """Return a flattened version of the ``source`` dict with dotted keys.
+
+        Example:
+            >>> event.source
+            {"content": {"body": "foo"}, "m.test": {"key": "bar"}}
+            >>> event.source.flattened()
+            {"content.body": "foo", "m.test.key": "bar"}
+
+        """
+
+        source = self.source if _source is None else _source
+        flat = {} if _flat is None else _flat
+
+        for key, value in source.items():
+            if isinstance(value, dict):
+                self.flattened(f"{_prefix}{key}.", value, flat)
+            else:
+                flat[f"{_prefix}{key}"] = value
+
+        return flat
+
+
+@dataclass
+class UnknownBadEvent(BaseEvent):
+    """An event that doesn't have the minimal necessary structure.
+
+    This type of event will be created if we can't find the event_id, sender,
+    origin server timestamp or event type.
+
+    The event can still be inspected with the source attribute.
+
+    """
+
 
 @dataclass
-class BadEvent:
+class BadEvent(BaseEvent):
```

Comment:
Given that this is a subclass of `BaseEvent` now, data fields (such as `source`) shouldn't be redeclared since they're now inherited